xarray/core/dataset.py

```diff
@@ -4028,36 +4028,20 @@ def unstack(
 
         result = self.copy(deep=False)
         for dim in dims:
-
             if (
-                # Dask arrays don't support assignment by index, which the fast unstack
-                # function requires.
-                # https://github.com/pydata/xarray/pull/4746#issuecomment-753282125
-                any(is_duck_dask_array(v.data) for v in self.variables.values())
-                # Sparse doesn't currently support (though we could special-case
-                # it)
-                # https://github.com/pydata/sparse/issues/422
-                or any(
-                    isinstance(v.data, sparse_array_type)
+                not sparse
+                # Dask arrays supports assignment by index,
+                # https://github.com/dask/dask/pull/7393
+                and all(
+                    dask_version >= "2021.04.0"
+                    and is_duck_dask_array(v.data)
+                    or isinstance(v.data, np.ndarray)
                     for v in self.variables.values()
                 )
-                or sparse
-                # Until https://github.com/pydata/xarray/pull/4751 is resolved,
-                # we check explicitly whether it's a numpy array. Once that is
-                # resolved, explicitly exclude pint arrays.
-                # # pint doesn't implement `np.full_like` in a way that's
-                # # currently compatible.
-                # # https://github.com/pydata/xarray/pull/4746#issuecomment-753425173
```

Comment:
(before we merge, let's add the comments that are still relevant — i.e. sparse & pint — back in....

Reply:
Done.